### PR TITLE
Implement date and relative-time formatters in ars-i18n

### LINKED ISSUE #125: task: Implement web-intl DateFormatter and RelativeTimeFormatter backend in ars-i18n
**Wave 4-6** | **Points: 5** | **Epic: #54 (I18n)**

## Goal

Implement the `web-intl` backend for date/time formatting in `ars-i18n` so WASM client builds can use browser `Intl.DateTimeFormat` and `Intl.RelativeTimeFormat` instead of ICU4X compiled data.

## Layer

Subsystem

## Framework

None

## Test tier

Unit

## Depends on

- #75 (ICU4X Locale integration)
- #129 (DateFormatter with ICU4X backend)

## Spec refs

- `spec/foundation/04-internationalization.md`
- `spec/shared/date-time-types.md`
- `spec/testing/14-ci.md`

## Files to create/modify

- `crates/ars-i18n/Cargo.toml`
- `crates/ars-i18n/src/lib.rs`
- `crates/ars-i18n/src/date.rs` and/or `crates/ars-i18n/src/web_intl/date.rs`
- `spec/foundation/04-internationalization.md`
- `spec/testing/14-ci.md` only if checks/spec drift need sync

## Tests to add first

- Feature-gated compile coverage proving `ars-i18n` builds with `--no-default-features --features web-intl --target wasm32-unknown-unknown`
- Tests for representative date formatting under locale-sensitive browser patterns
- Tests for relative time formatting via browser `Intl.RelativeTimeFormat`
- Tests proving `icu4x` and `web-intl` are mutually exclusive

## Acceptance criteria

- `web-intl` is a real backend for the public date/time formatter surface.
- Browser builds use `Intl.DateTimeFormat` for date/time formatting where the spec maps cleanly to browser capabilities.
- Browser builds use `Intl.RelativeTimeFormat` for relative time output.
- The public `DateFormatter` / related API stays stable across `icu4x` and `web-intl` builds.
- Unsupported browser gaps are handled explicitly in code and spec rather than silently changing semantics.
- CI/verification includes the wasm `web-intl` cargo check path required by the spec.

## Spec impact

Likely yes.

If browser APIs cannot exactly match every ICU4X-oriented option or calendar behavior in the current spec, document the supported subset and fallback behavior in the same task.

### LINKED ISSUE #129: task: Implement DateFormatter with ICU4X backend in ars-i18n
**Wave 4a** | **Points: 3** | **Epic: #54 (I18n)**

## Goal

Implement locale-aware date formatting using ICU4X `DateTimeFormatter` with field sets.

## Layer

Subsystem

## Framework

None

## Test tier

Unit

## Depends on

- #75 (ICU4X Locale integration)
- #128 (CalendarDate internal type)

## Spec refs

- `spec/foundation/04-internationalization.md` §5.4

## Scope

- `FormatLength` enum (Full, Long, Medium, Short) with `to_icu_date_field_set()`, `to_icu_full_date_field_set()`, `is_full()`, `to_icu_time_field_set()`
- `DateFormatterInner` enum with `Ymd(DateTimeFormatter<YMD>)`, `Ymde(DateTimeFormatter<YMDE>)`
- `DateFormatter` struct with `new(locale, length)` and `format(date)`

## Tests to add first

- `DateFormatter::new(&en_us, FormatLength::Short).format(date)` produces locale-correct short date
- `FormatLength::Full` includes weekday name
- Non-Gregorian calendars format correctly (e.g., Japanese era year)
- Multiple locales: en-US, de-DE, ar-SA, ja-JP

## Acceptance criteria

- `DateFormatter` backed by ICU4X `DateTimeFormatter` with YMD/YMDE field sets
- All four format lengths produce correct output
- Non-Gregorian calendar dates format correctly
- All tests TDD-first, zero warnings

### LINKED ISSUE #130: task: Implement RelativeTimeFormatter with ICU4X backend in ars-i18n
**Wave 4a** | **Points: 3** | **Epic: #54 (I18n)**

## Goal

Implement relative time formatting ("2 days ago", "in 1 hour") using `icu_experimental::relativetime`.

## Layer

Subsystem

## Framework

None

## Test tier

Unit

## Depends on

- #75 (ICU4X Locale integration)

## Spec refs

- `spec/foundation/04-internationalization.md` §5.5

## Scope

- `NumericOption` enum (Always, Auto) with `to_icu()`
- `RelativeTimeFormatter` struct with `new()`, `with_numeric()`, `format_seconds()`
- Uses `icu_experimental::relativetime` (NOT deprecated `icu_relativetime` crate)

## Tests to add first

- `format_seconds(-60)` returns "1 minute ago" (en-US)
- `format_seconds(3600)` returns "in 1 hour" (en-US)
- `NumericOption::Auto` returns "yesterday" for -86400s (en-US)
- Arabic locale produces Arabic output
- Multiple time unit boundaries (seconds, minutes, hours, days)

## Acceptance criteria

- `RelativeTimeFormatter` backed by `icu_experimental::relativetime`
- Correct time unit decomposition from seconds
- `NumericOption::Auto` produces natural language ("yesterday") where applicable
- All tests TDD-first, zero warnings

### PR DESCRIPTION
**Pull request by fogodev**
## Summary
- add backend-neutral `DateFormatter` and `RelativeTimeFormatter` to `ars-i18n`
- implement ICU4X and `web-intl` formatter backends, including browser-backed relative time and date formatting
- refactor web date conversion to avoid `icu/compiled_data`, keep calendar parity through the internal calendar bridge, and sync the i18n/date-time specs

## Verification
- `cargo test -p ars-i18n`
- `cargo test -p ars-i18n --no-default-features --features web-intl --target wasm32-unknown-unknown --no-run`
- `CARGO_TARGET_WASM32_UNKNOWN_UNKNOWN_RUNNER=wasm-bindgen-test-runner cargo test -p ars-i18n --no-default-features --features web-intl --target wasm32-unknown-unknown`
- `cargo xci` (run from `/private/tmp/ars-ui-07d0` to avoid the macOS wasm coverage linker path-length issue in the worktree path)

Closes #125
Closes #129
Closes #130

### CODE REVIEW

File: crates/ars-i18n/src/date.rs [hunk 1]
```diff
@@ -0,0 +1,600 @@
...
+        .expect("CalendarDate should be valid for browser date formatting");
+
+    let gregorian = internal.to_calendar(CalendarSystem::Gregorian);
+
+    js_date_from_ymd(gregorian.year(), gregorian.month(), gregorian.day())
+}
+
+#[cfg(all(feature = "web-intl", target_arch = "wasm32", not(feature = "icu4x")))]
+fn js_date_from_ymd(year: i32, month: u8, day: u8) -> JsDate {
+    let millis = JsDate::utc(f64::from(year), f64::from(month - 1))
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-i18n/src/date.rs]: Fixed in `052d5ea`. `js_date_from_ymd` now starts from epoch `0` and sets the UTC year/month/day explicitly, then pins the time to `12:00 UTC`, which avoids the JavaScript `1900..1999` remapping for years `0..99` while keeping the helper timezone-neutral. I also added a browser-executed wasm regression test, `web_intl_date_helpers_preserve_low_gregorian_years`, that asserts both the raw helper and `js_date_from_calendar` preserve year `44` exactly.

File: crates/ars-i18n/src/date.rs [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+
+    js_date_from_ymd(gregorian.year(), gregorian.month(), gregorian.day())
+}
+
+#[cfg(all(feature = "web-intl", target_arch = "wasm32", not(feature = "icu4x")))]
+fn js_date_from_ymd(year: i32, month: u8, day: u8) -> JsDate {
+    let date = JsDate::new(&JsValue::from_f64(0.0));
+
+    date.set_utc_full_year_with_month_date(
+        u32::try_from(year).expect("browser-backed date formatting requires Gregorian years >= 0"),
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-i18n/src/date.rs]: Fixed in `bb00f6d`. `js_date_from_ymd` no longer routes years through `u32::try_from`; it now builds the UTC date from a canonical ISO-8601 string, so signed Gregorian years are preserved for the browser-backed path. I also added a browser-executed wasm regression test, `web_intl_date_helpers_support_pre_ce_gregorian_dates`, that locks the helper on negative, zero, low, and 5-digit year encodings (`-1`, `0`, `44`, `10000`). I checked for a clean public `CalendarDate` fixture that actually reaches BCE through the current provider conversions and did not find one in the supported ranges, so the regression is targeted at the helper boundary where the panic lived.

File: crates/ars-i18n/src/date.rs [hunk 3]
```diff
@@ -0,0 +1,642 @@
...
+            length,
+            inner,
+        }
+    }
+
+    /// Formats a calendar date for the formatter locale.
+    #[must_use]
+    pub fn format(&self, date: &CalendarDate) -> String {
+        let internal = InternalCalendarDate::try_from(date)
+            .expect("CalendarDate should be valid for formatting");
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-i18n/src/date.rs]: Fixed in `91793f5`. Both formatter backends now return a deterministic fallback string instead of panicking when `InternalCalendarDate::try_from` rejects an otherwise valid public `CalendarDate` such as Gregorian year `10000`. On the browser side I also bypass the internal conversion entirely for Gregorian input, so `web-intl` can still hand `10000-01-01` directly to `Intl.DateTimeFormat`. I added regressions for both paths: `formatter_gracefully_formats_large_gregorian_years` under ICU4X and `web_intl_date_formatter_gracefully_formats_large_gregorian_years` under the wasm `web-intl` backend.

File: crates/ars-i18n/src/date.rs [hunk 4]
```diff
@@ -0,0 +1,646 @@
...
+    pub fn format(&self, date: &CalendarDate) -> String {
+        let Some(js_date) = js_date_from_calendar(date) else {
+            return fallback_format_date(&self.locale, self.length, date);
+        };
+
+        let format: Function = self.inner.format();
+
+        format
+            .call1(&JsValue::UNDEFINED, js_date.as_ref())
+            .expect("Intl.DateTimeFormat.format should not throw for a valid Date")
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-i18n/src/date.rs:208]: Fixed in `c25fc40`. The wasm `web-intl` path now treats invalid JavaScript `Date` values as a conversion miss by validating the constructed `Date` with `get_time().is_nan()` inside `js_date_from_calendar`. That means out-of-range public Gregorian inputs such as year `1_000_000` now return the deterministic fallback string instead of reaching `Intl.DateTimeFormat.format` and panicking on a `RangeError`. I also added a browser-executed regression, `web_intl_date_formatter_falls_back_for_js_out_of_range_gregorian_years`, which checks both the invalid-date helper path and the formatter fallback.

File: crates/ars-i18n/src/relative_time.rs [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+                .expect("compiled_data guarantees hour relative-time data"),
+            day_formatter: IcuRelativeTimeFormatter::try_new_long_day(prefs, options)
+                .expect("compiled_data guarantees day relative-time data"),
+        }
+    }
+
+    /// Formats a duration in seconds relative to now.
+    #[must_use]
+    pub fn format_seconds(&self, seconds: i64) -> String {
+        if seconds.abs() < 60 {
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-i18n/src/relative_time.rs]: Fixed in `d64ec6a`. Relative-time bucketing now goes through a shared `bucket_relative_time` helper built on `unsigned_abs()`, so `i64::MIN` no longer overflows when selecting the second/minute/hour/day formatter path. I also added regressions that lock the edge explicitly on both sides: `bucketing_handles_i64_min_without_overflow` for the native path and `web_intl_bucketing_handles_i64_min_without_overflow` for the browser-backed wasm path.

File: crates/ars-i18n/src/lib.rs [hunk 6]
```diff
@@ -378,6 +375,30 @@ pub trait IcuProvider: Send + Sync + 'static {
             return date.clone();
         }
 
+        #[cfg(any(feature = "icu4x", feature = "web-intl"))]
+        {
+            let internal = calendar::internal::CalendarDate::try_from(date)
+                .expect("CalendarDate should be valid for calendar conversion");
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-i18n/src/lib.rs]: Fixed in `b2a44a8`. `StubIcuProvider::convert_date` now treats internal bridge rejection as a deterministic fallback instead of unwrapping: if `calendar::internal::CalendarDate::try_from(date)` fails for an otherwise valid public input such as Gregorian year `10000`, the provider now returns the original `CalendarDate` unchanged. I also added `stub_icu_provider_convert_date_falls_back_for_inputs_outside_internal_bridge_range`, which locks both the direct provider call and `CalendarDate::to_calendar()` on that boundary.

File: crates/ars-i18n/src/lib.rs [hunk 7]
```diff
@@ -378,6 +375,31 @@ pub trait IcuProvider: Send + Sync + 'static {
...
+                return date.clone();
+            };
+            let converted = internal.to_calendar(target);
+
+            CalendarDate {
+                calendar: target,
+                era: converted
+                    .era()
+                    .filter(|_| target.has_custom_eras())
+                    .map(|code| Era {
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-i18n/src/lib.rs:390]: Fixed in 62cffb6.

The shared internal calendar bridge now preserves astronomical year numbering when materializing a public Gregorian `CalendarDate`, instead of always using ICU era-year output. That fixes the provider conversion path and the `web-intl` non-Gregorian formatting path, since the browser formatter also converts through the same internal Gregorian bridge.

I also added regressions for both layers:
- `stub_icu_provider_convert_date_preserves_bce_gregorian_years`
- `web_intl_date_helpers_preserve_astronomical_gregorian_years_from_non_gregorian_input`

The same commit also includes the extra `date.rs` / `relative_time.rs` coverage work we discussed, and `cargo xci` is green on the updated branch.
